added has X role functions

ciaranRoche opened the pull request:
## Motivation

JIRA: https://issues.jboss.org/browse/AGDROID-685

## Description

Added two functions to check if a user has a given realm role and a given client role.

## Progress

- [x] Implementation
- [x] Java Doc
- [x] Unit Test

ciaranRoche commented:
@rachael-oregan @aidenkeating @ziccardi

aidenkeating commented on auth/src/main/java/org/aerogear/auth/impl/OIDCAuthCodeImpl.java:
```diff
@@ -129,6 +129,47 @@ private String getEmail() throws JSONException {
+        boolean hasRole = false;
+        try{
+            List<RealmRole> realmRoles = getRealmRoles();
+            if (realmRoles.contains(role)){
```
I wouldn't have thought this would work? Trying to obtain a `String` from a list of `RealmRole`?

ciaranRoche replied:
Good spot, ill change that. Thanks.

aidenkeating replied:
@rachael-oregan I think if we refactor `getRealmRoles` and `getClientRoles` to not do the parsing and instead do it "on build" these functions can become a lot simpler. So we don't have to handle a JSONException. Something like:

```java
public boolean hasRealmRole(String roleName) {
  return realmRoles.stream().filter(role -> role.getName().equals(roleName)).findFirst().isPresent();
}

...
```

aidenkeating commented on auth/src/main/java/org/aerogear/auth/impl/OIDCAuthCodeImpl.java:
```diff
@@ -129,6 +129,47 @@ private String getEmail() throws JSONException {
+        boolean hasRole = false;
+        try {
+            List<ClientRole> clientRoles = getClientRoles();
+            if( clientRoles.contains(role)){
```
See other comment about the list of `RealmRole`. I think the same applies here.

ziccardi commented on auth/src/main/java/org/aerogear/auth/impl/UserPrincipalImpl.java:
```diff
@@ -119,8 +121,40 @@ UserPrincipalImpl build() {
+     * @return true or false
+     */
+    @Override
+    public boolean hasClientRole(String role){
```
The role parameter should be final

ziccardi replied:
If for some reason in the future we will need to have another type of roles, this way we will have to add new methods. I think it would be better to change this method to 
`hasRole(final String roleName, final Type roleType)`
where `roleType` is an enum.

ziccardi commented on auth/src/main/java/org/aerogear/auth/impl/UserPrincipalImpl.java:
```diff
@@ -119,8 +121,40 @@ UserPrincipalImpl build() {
+     * @return true or false
+     */
+    @Override
+    public boolean hasRealmRole(String role){
```
The role parameter should be final

ziccardi commented on auth/src/main/java/org/aerogear/auth/impl/UserPrincipalImpl.java:
```diff
@@ -119,8 +121,40 @@ UserPrincipalImpl build() {
+    @Override
+    public boolean hasClientRole(String role){
+        for ( IRole r : roles){
+            if (r.getRoleName().equals(role) && r instanceof ClientRole){
```
This way, every time we ask if a role is present, in the worst case the whole list of roles is fetched.
Wouldn't be better to save the roles in a Map so that the `containsKey` method could be used?

ziccardi replied:
Another issue: are role names case sensitive? The standard implementation of the `equals` method is case sensitive.

ziccardi commented on auth/src/main/java/org/aerogear/auth/impl/UserPrincipalImpl.java:
```diff
@@ -119,8 +121,40 @@ UserPrincipalImpl build() {
+    @Override
+    public boolean hasRealmRole(String role){
+        for ( IRole r : roles){
+            if ( r.getRoleName().equals(role) && r instanceof RealmRole){
```
This way, every time we ask if a role is present, in the worst case the whole list of roles is fetched.
Wouldn't be better to save the roles in a `Map` so that the `containsKey` method could be used?

ziccardi replied:
Another issue: are role names case sensitive? The standard implementation of the `equals` method is case sensitive.

ciaranRoche commented:
@ziccardi @aidenkeating @rachael-oregan 
Hey guys could you have a look at my latest commit. I added testing to test the hasRole() method, but could not get them to pass, on debug, the containsKey() does not call the override equals method in the RoleKey class.
I added a different hasRole() method that iterates over the Map and with this it calls the override equals method, and the tests pass.
So, we are thinking, this is down to the unit tests and android sdk. As it just goes to a containsKey stub and returns false. 
If you can give the thumbs up that the containsKey() will work at run time and we could change the tests to test the equals method, instead of directly testing the hasRole() method which essentially is just a wrapper to the containsKey() method.

ziccardi commented:
@ciaranRoche implement the hashcode as well

rachael-oregan commented:
@ciaranRoche I merged this into my branch [here](https://github.com/aerogear/aerogear-android-sdk/pull/15/files) so going to close this

ciaranRoche commented:
Cheers Rachael :+1:
